File: frontend/src/services/admin/placementTestManagementService.ts
import apiClient from "@/config/api.ts";

export interface PlacementTestAttempt {
    id: number;
    status: "IN_PROGRESS" | "COMPLETED";
    totalScore: number | null;
    detectedLevelName: string | null;
    createdAt: string;
    completedAt: string | null;
    duration: string | null;
}

export interface PlacementTestStats {
    totalTests: number;
    completedTests: number;
    inProgressTests: number;
    averageScore: number;
}

export interface PlacementTestRecord {
    id: number;
    userId: number;
    userName: string;
    userEmail: string;
    userAvatar: string;
    status: "IN_PROGRESS" | "COMPLETED";
    totalScore: number | null;
    detectedLevel: string | null;
    createdAt: string;
    completedAt: string | null;
    /** Thời lượng làm bài, ví dụ "12p 34s" hoặc "1g 02p 15s" */
    duration: string | null;
    /** Tổng số lần user đã làm placement test */
    totalAttempts: number;
}

interface PageResponse<T> {
    content: T[];
    totalElements: number;
    totalPages: number;
    number: number;
    size: number;
}

interface BackendPlacementTest {
    id: number;
    userId: number;
    userName: string;
    userEmail: string;
    userAvatar: string | null;
    status: string;
    totalScore: number | null;
    detectedLevelName: string | null;
    createdAt: string;
    completedAt: string | null;
    totalAttempts: number;
}

/**
 * Chuyển đổi dữ liệu placement test từ backend sang định dạng UI
 */
function mapPlacementTest(t: BackendPlacementTest): PlacementTestRecord {
    function formatDatetime(raw: string | null): string | null {
        if (!raw) return null;
        const d = new Date(raw);
        const date = d.toLocaleDateString("vi-VN", { day: "2-digit", month: "2-digit", year: "numeric" });
        const time = d.toLocaleTimeString("vi-VN", { hour: "2-digit", minute: "2-digit", second: "2-digit", hour12: false });
        return `${date} ${time}`;
    }

    const createdAt  = formatDatetime(t.createdAt)  ?? "N/A";
    const completedAt = formatDatetime(t.completedAt);

    // Tính thời lượng làm bài (createdAt → completedAt)
    let duration: string | null = null;
    if (t.createdAt && t.completedAt) {
        const diffMs = new Date(t.completedAt).getTime() - new Date(t.createdAt).getTime();
        if (diffMs > 0) {
            const totalSec = Math.floor(diffMs / 1000);
            const h = Math.floor(totalSec / 3600);
            const m = Math.floor((totalSec % 3600) / 60);
            const s = totalSec % 60;
            duration = h > 0
                ? `${h}g ${String(m).padStart(2, "0")}p ${String(s).padStart(2, "0")}s`
                : `${String(m).padStart(2, "0")}p ${String(s).padStart(2, "0")}s`;
        }
    }

    return {
        id: t.id,
        userId: t.userId,
        userName: t.userName || t.userEmail.split("@")[0],
        userEmail: t.userEmail,
        userAvatar: t.userAvatar || `https://ui-avatars.com/api/?name=${encodeURIComponent(t.userName || t.userEmail)}&background=f97316&color=fff`,
        status: t.status as PlacementTestRecord["status"],
        totalScore: t.totalScore,
        detectedLevel: t.detectedLevelName,
        createdAt,
        completedAt,
        duration,
        totalAttempts: t.totalAttempts ?? 0,
    };
}

export const placementTestManagementService = {
    /**
     * Lấy danh sách placement tests có phân trang
     * GET /api/admin/placement-tests?page={page}&size={size}
     */
    async getTests(page = 0, size = 10): Promise<{ tests: PlacementTestRecord[]; total: number }> {
        const res = await apiClient.get<{ data: PageResponse<BackendPlacementTest> }>(
            `/admin/placement-tests?page=${page}&size=${size}`
        );
        const data = res.data.data;
        return {
            tests: data.content.map(mapPlacementTest),
            total: data.totalElements,
        };
    },

    /**
     * Lấy thống kê placement tests
     * GET /api/admin/placement-tests/stats
     */
    async getStats(): Promise<PlacementTestStats> {
        const res = await apiClient.get<{ data: PlacementTestStats }>("/admin/placement-tests/stats");
        return res.data.data;
    },

    /**
     * Lấy chi tiết một placement test
     * GET /api/admin/placement-tests/{id}
     */
    async getTestDetail(id: number): Promise<PlacementTestRecord> {
        const res = await apiClient.get<{ data: BackendPlacementTest }>(`/admin/placement-tests/${id}`);
        return mapPlacementTest(res.data.data);
    },

    /**
     * Xóa một placement test
     * DELETE /api/admin/placement-tests/{id}
     */
    async deleteTest(id: number): Promise<void> {
        await apiClient.delete(`/admin/placement-tests/${id}`);
    },

    /**
     * Lịch sử tất cả các lần làm bài của một user
     * GET /api/admin/placement-tests/user/{userId}/history
     */
    async getUserHistory(userId: number): Promise<PlacementTestAttempt[]> {
        const res = await apiClient.get<{ data: Array<{
            id: number;
            status: string;
            totalScore: number | null;
            detectedLevelName: string | null;
            createdAt: string;
            completedAt: string | null;
        }> }>(`/admin/placement-tests/user/${userId}/history`);

        return res.data.data.map(t => {
            const createdAt = t.createdAt
                ? (() => {
                    const d = new Date(t.createdAt);
                    return `${d.toLocaleDateString("vi-VN", { day: "2-digit", month: "2-digit", year: "numeric" })} ${d.toLocaleTimeString("vi-VN", { hour: "2-digit", minute: "2-digit", second: "2-digit", hour12: false })}`;
                })()
                : "N/A";
            const completedAt = t.completedAt
                ? (() => {
                    const d = new Date(t.completedAt!);
                    return `${d.toLocaleDateString("vi-VN", { day: "2-digit", month: "2-digit", year: "numeric" })} ${d.toLocaleTimeString("vi-VN", { hour: "2-digit", minute: "2-digit", second: "2-digit", hour12: false })}`;
                })()
                : null;
            let duration: string | null = null;
            if (t.createdAt && t.completedAt) {
                const diffMs = new Date(t.completedAt).getTime() - new Date(t.createdAt).getTime();
                if (diffMs > 0) {
                    const totalSec = Math.floor(diffMs / 1000);
                    const h = Math.floor(totalSec / 3600);
                    const m = Math.floor((totalSec % 3600) / 60);
                    const s = totalSec % 60;
                    duration = h > 0
                        ? `${h}g ${String(m).padStart(2, "0")}p ${String(s).padStart(2, "0")}s`
                        : `${String(m).padStart(2, "0")}p ${String(s).padStart(2, "0")}s`;
                }
            }
            return {
                id: t.id,
                status: t.status as PlacementTestAttempt["status"],
                totalScore: t.totalScore,
                detectedLevelName: t.detectedLevelName,
                createdAt,
                completedAt,
                duration,
            };
        });
    },
};

File: backend/src/main/java/com/languagelearning/service/admin/PlacementTestManagementService.java
package com.languagelearning.service.admin;

import com.languagelearning.dto.admin.placement_test_management.PlacementTestDto;
import com.languagelearning.dto.admin.placement_test_management.PlacementTestStatsDto;
import com.languagelearning.dto.admin.placement_test_management.PlacementTestAttemptDto;
import com.languagelearning.entity.PlacementTest;
import com.languagelearning.entity.User;
import com.languagelearning.entity.UserProfile;
import com.languagelearning.repository.mysql.PlacementTestRepository;
import com.languagelearning.repository.mysql.UserProfileRepository;
import com.languagelearning.repository.mysql.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class PlacementTestManagementService {

    private final PlacementTestRepository placementTestRepository;
    private final UserProfileRepository userProfileRepository;
    private final UserRepository userRepository;

    /**
     * Lấy danh sách placement tests — mỗi user chỉ hiển thị lần thi MỚI NHẤT,
     * sắp xếp theo createdAt DESC, có phân trang.
     */
    @Transactional(readOnly = true)
    public Page<PlacementTestDto> getTests(int page, int size) {
        // Lấy bản ghi mới nhất của mỗi user
        List<PlacementTest> latest = placementTestRepository.findLatestPerUser();

        // Phân trang thủ công
        int start = page * size;
        int end   = Math.min(start + size, latest.size());
        List<PlacementTest> pageContent = start < latest.size()
                ? latest.subList(start, end)
                : List.of();

        List<PlacementTestDto> dtos = pageContent.stream()
                .map(this::toDto)
                .toList();

        return new PageImpl<>(dtos, PageRequest.of(page, size), latest.size());
    }

    /**
     * Thống kê tổng quan:
     * - totalTests      : tổng số bản ghi (mọi lần thi)
     * - completedTests  : số bản ghi COMPLETED
     * - inProgressTests : số bản ghi IN_PROGRESS (chưa hoàn thành / bỏ dở)
     * - averageScore    : điểm TB của các bài COMPLETED
     */
    @Transactional(readOnly = true)
    public PlacementTestStatsDto getStats() {
        List<PlacementTest> all = placementTestRepository.findAll();
        long total      = all.size();
        long completed  = all.stream().filter(t -> "COMPLETED".equals(t.getStatus())).count();
        long inProgress = all.stream().filter(t -> "IN_PROGRESS".equals(t.getStatus())).count();

        double averageScore = all.stream()
                .filter(t -> "COMPLETED".equals(t.getStatus()) && t.getTotalScore() != null)
                .mapToDouble(PlacementTest::getTotalScore)
                .average()
                .orElse(0.0);

        return PlacementTestStatsDto.builder()
                .totalTests(total)
                .completedTests(completed)
                .inProgressTests(inProgress)
                .averageScore(averageScore)
                .build();
    }

    /**
     * Lấy chi tiết một placement test theo ID.
     */
    @Transactional(readOnly = true)
    public PlacementTestDto getTestDetail(Integer testId) {
        PlacementTest test = placementTestRepository.findById(testId)
                .orElseThrow(() -> new IllegalArgumentException("Placement test not found: " + testId));
        return toDto(test);
    }

    /**
     * Xóa một placement test theo ID.
     */
    @Transactional
    public void deleteTest(Integer testId) {
        PlacementTest test = placementTestRepository.findById(testId)
                .orElseThrow(() -> new IllegalArgumentException("Placement test not found: " + testId));
        placementTestRepository.delete(test);
    }

    /**
     * Lịch sử tất cả các lần làm bài của một user, mới nhất trước.
     */
    @Transactional(readOnly = true)
    public List<PlacementTestAttemptDto> getUserHistory(Integer userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + userId));
        return placementTestRepository.findByUserIdOrderByCreatedAtDesc(user.getId())
                .stream()
                .map(t -> {
                    LocalDateTime completedAt = "COMPLETED".equals(t.getStatus()) ? t.getUpdatedAt() : null;
                    String levelName = null;
                    if (t.getDetectedLevel() != null) {
                        levelName = switch (t.getDetectedLevel().getId()) {
                            case 1 -> "Beginner";
                            case 2 -> "Intermediate";
                            case 3 -> "Advanced";
                            default -> t.getDetectedLevel().getLevelName();
                        };
                    }
                    return PlacementTestAttemptDto.builder()
                            .id(t.getId())
                            .status(t.getStatus())
                            .totalScore(t.getTotalScore())
                            .detectedLevelName(levelName)
                            .createdAt(t.getCreatedAt())
                            .completedAt(completedAt)
                            .build();
                })
                .toList();
    }

    // helpers

    private PlacementTestDto toDto(PlacementTest test) {
        Optional<UserProfile> profileOpt = userProfileRepository.findByUserId(test.getUser().getId());

        String userName = profileOpt.map(UserProfile::getFullName).orElse(null);
        if (userName == null || userName.isBlank()) {
            userName = test.getUser().getEmail().split("@")[0];
        }

        String avatarUrl = profileOpt.map(UserProfile::getAvatarUrl).orElse(null);
        if (avatarUrl == null || avatarUrl.isBlank()) {
            avatarUrl = "https://ui-avatars.com/api/?name=" + userName + "&background=f97316&color=fff";
        }

        String detectedLevelName = null;
        if (test.getDetectedLevel() != null) {
            // Map theo id vì level_name trong DB có thể khác với giá trị frontend dùng để filter
            detectedLevelName = switch (test.getDetectedLevel().getId()) {
                case 1 -> "Beginner";
                case 2 -> "Intermediate";
                case 3 -> "Advanced";
                default -> test.getDetectedLevel().getLevelName();
            };
        }

        // completedAt: chỉ có khi COMPLETED
        LocalDateTime completedAt = "COMPLETED".equals(test.getStatus()) ? test.getUpdatedAt() : null;

        // Tổng số lần user đã làm (cả hoàn thành lẫn bỏ dở)
        long totalAttempts = placementTestRepository.countDistinctByUserId(test.getUser().getId());

        return PlacementTestDto.builder()
                .id(test.getId())
                .userId(test.getUser().getId())
                .userName(userName)
                .userEmail(test.getUser().getEmail())
                .userAvatar(avatarUrl)
                .status(test.getStatus())
                .totalScore(test.getTotalScore())
                .detectedLevelName(detectedLevelName)
                .createdAt(test.getCreatedAt())
                .completedAt(completedAt)
                .totalAttempts(totalAttempts)
                .build();
    }
}
